### PR TITLE
Add assertions for assignment versions of fundamental operators + symmetric difference

### PR DESCRIPTION
**Pull request by saik0**
This PR adds proptests for assigning operators, plus tests for symmetric difference

### CODE REVIEW

**Comment by saik0** [comment]
Clippy fix fixed the "unnecessary" taking references. Closing until I put them back in.

**Comment by saik0** [comment]
@Kerollmops Lets get this merged before SIMD so we can gave the ops properly tested

File: src/bitmap/proptests.rs [hunk 1]
```diff
@@ -528,7 +528,11 @@ mod test {
             prop_assert_eq!(&a, &b);
             prop_assert!(a.is_subset(&b) && b.is_subset(&a));
 
-            b.insert(a.max().unwrap_or(0) + 1);
```
Comment (saik0) [inline src/bitmap/proptests.rs:164]: This will panic if `max() == u32::MAX`

Comment (Kerollmops) [reply src/bitmap/proptests.rs:164]: You are absolutely right!

Comment (saik0) [reply src/bitmap/proptests.rs:164]: I know. proptest told me so. In big red letters.

File: src/bitmap/proptests.rs [hunk 2]
```diff
@@ -234,16 +608,130 @@ mod test {
...
+                c_sub_a -= a.clone();
+
+                let mut c_sub_b = c.clone();
+                c_sub_b -= b.clone();
+
+                let mut x = c.clone();
+                x -= a_or_b.clone();
+
+                let mut y = c_sub_a;
+                y &= c_sub_b.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:687]:
```suggestion
                let mut c_sub_b = c.clone();
                c_sub_b -= b;

                let mut x = c.clone();
                x -= a_or_b;

                let mut y = c_sub_a;
                y &= c_sub_b;
```

File: src/bitmap/proptests.rs [hunk 3]
```diff
@@ -234,16 +608,130 @@ mod test {
...
+                c_sub_a -= a.clone();
+
+                let mut c_sub_b = c.clone();
+                c_sub_b -= b.clone();
+
+                let mut x = c.clone();
+                x -= a_and_b.clone();
+
+                let mut y = c_sub_a;
+                y |= c_sub_b.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:644]:
```suggestion
                let mut c_sub_b = c.clone();
                c_sub_b -= b;

                let mut x = c.clone();
                x -= a_and_b;

                let mut y = c_sub_a;
                y |= c_sub_b;
```

File: src/bitmap/proptests.rs [hunk 4]
```diff
@@ -234,16 +608,130 @@ mod test {
...
+                a_and_c &= c.clone();
+
+                let mut c_sub_b = c.clone();
+                c_sub_b -= b.clone();
+
+                let mut x = c.clone();
+                x -= b_sub_a.clone();
+
+                let mut y = a_and_c;
+                y |= c_sub_b.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:730]:
```suggestion
                let mut c_sub_b = c.clone();
                c_sub_b -= b;

                let mut x = c.clone();
                x -= b_sub_a;

                let mut y = a_and_c;
                y |= c_sub_b;
```

File: src/bitmap/proptests.rs [hunk 5]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                c_sub_a -= a.clone();
+
+                let mut x = b_sub_a;
+                x &= c.clone();
+
+                let mut y = b_and_c;
+                y -= a.clone();
+
+                let mut z = c_sub_a;
+                z &= b.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:786]:
```suggestion
                let mut x = b_sub_a;
                x &= c;

                let mut y = b_and_c;
                y -= a;

                let mut z = c_sub_a;
                z &= b;
```

File: src/bitmap/proptests.rs [hunk 6]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                b_or_c |= c.clone();
+
+                let mut a_sub_c = a.clone();
+                a_sub_c -= c.clone();
+
+                let mut x = b_sub_a;
+                x |= c.clone();
+
+                let mut y = b_or_c;
+                y -= a_sub_c.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:831]:
```suggestion
                let mut x = b_sub_a;
                x |= c;

                let mut y = b_or_c;
                y -= a_sub_c;
```

File: src/bitmap/proptests.rs [hunk 7]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                b_sub_a -= a.clone();
+
+                let mut a_or_c = a.clone();
+                a_or_c |= c.clone();
+
+                let mut x = b_sub_a;
+                x -= c;
+
+                let mut y = b;
+                y -= a_or_c.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:868]:
```suggestion
                y -= a_or_c;
```

File: src/bitmap/proptests.rs [hunk 8]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+            { // op assign ref
+                let mut x = a.clone();
+                x -= &a;
+
+                prop_assert_eq!(x, empty_set());
+            }
+
+            { // op assign own
+                let mut x = a.clone();
+                x -= a.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:887]:
```suggestion
                x -= a;
```

File: src/bitmap/proptests.rs [hunk 9]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+            { // op assign ref
+            let mut x = empty_set();
+            x -= &a;
+
+            prop_assert_eq!(x, empty_set());
+        }
+
+        { // op assign own
+                let mut x = empty_set();
+                x -= a.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:906]:
```suggestion
                x -= a;
```

File: src/bitmap/proptests.rs [hunk 10]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+            );
+
+            { // op assign ref
+                let mut a_xor_b = a.clone();
+                a_xor_b ^= &b;
+
+                let mut b_xor_c = b.clone();
+                b_xor_c ^= &c;
+
+                let mut a_xor_c = a.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:955]:
```suggestion
                let mut b_xor_c = b;
                b_xor_c ^= &c;

                let mut a_xor_c = a;
```

File: src/bitmap/proptests.rs [hunk 11]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                a_xor_b ^= b.clone();
+
+                let mut b_xor_c = b;
+                b_xor_c ^= c.clone();
+
+                let mut a_xor_c = a;
+                a_xor_c ^= c;
+
+                let mut tri = a_xor_b;
+                tri ^= b_xor_c.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:975]:
```suggestion
                tri ^= b_xor_c;
```

**Comment by saik0** [comment]
@Kerollmops SIMD PR is ready to go once the rest are merged. 🎉

File: src/bitmap/proptests.rs [hunk 12]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                x ^= b.clone();
+
+                let mut a_sub_b = a.clone();
+                a_sub_b -= b.clone();
+
+                let mut b_sub_a = b.clone();
+                b_sub_a -= a.clone();
+
+                let mut y = a_sub_b;
+                y |= b_sub_a.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:1068]:
```suggestion
                let mut b_sub_a = b.clone();
                b_sub_a -= a;

                let mut y = a_sub_b;
                y |= b_sub_a;
```

File: src/bitmap/proptests.rs [hunk 13]
```diff
@@ -254,30 +742,370 @@ mod test {
...
+                x ^= b.clone();
+
+                let mut a_or_b = a.clone();
+                a_or_b |= b.clone();
+
+                let mut a_and_b = a;
+                a_and_b &= b;
+
+                let mut y = a_or_b;
+                y -= a_and_b.clone();
```
Comment (Kerollmops) [inline src/bitmap/proptests.rs:1105]:
```suggestion
                y -= a_and_b;
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Not sure if my review is useful at all, as performance doesn't matter much in the tests. However, isn't Clippy able to find out that some clones are useless?

**Comment by saik0** [comment]
Not all of the suggested changes were useless clones, I removed the ones I could. I'm not sure why clippy isn't catching this.

**Review by Kerollmops — APPROVED** [review]
If there are more useless clones, that's not very important so let's merge!
bors merge

**Pull request merged**